### DISCUSSION TITLE
Moving around in scene, reference space error

### DISCUSSION THREAD

**Discussion by paulmelis** [StereoKit/StereoKit#1041] [Q&A]
I'm trying to make a simple way to move around in a 3D scene. I'm doing this by changing `World.OriginOffset`, but I'm wondering if that's the best way. Based on some controller input I update the value:

```
Controller rc = Input.Controller(Handed.Right);

if (!trigger_active && rc.trigger > 0.5f)
{
    Vec3 dir = rc.pose.Forward.Normalized;
    dir.y = -dir.y;
    World.OriginOffset = new Pose(World.OriginOffset.position + dir, World.OriginOffset.orientation);
    trigger_active = true;
}
else if (rc.trigger < 0.5f)
    trigger_active = false;

int stick = 0;
if (rc.stick.x > 0.5f)
    stick = 1;
else if (rc.stick.x < -0.5f)
    stick = -1;

if (!stick_active && stick != 0)
{
    stick_active = true;
    World.OriginOffset = new Pose(World.OriginOffset.position, World.OriginOffset.orientation * Quat.FromAngles(0f, -stick*45f, 0f));
    trigger_active = true;
}
else if (stick == 0)
    stick_active= false;
```

However, this seems to crash the app, apparently when doing too many rotations. I get this:

```
[SK info] xrCreateReferenceSpace failed [XR_ERROR_POSE_INVALID]
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Repeat 2 times:
--------------------------------
   at StereoKit.NativeAPI.sk_step(System.Action)
--------------------------------
   at StereoKit.SK.Step(System.Action)
   at StereoKit.SK.Run(System.Action, System.Action)
   at Program.<Main>$(System.String[])
```

![image](https://github.com/StereoKit/StereoKit/assets/1408287/26995f46-e5d3-4001-9c77-e12d7e82a0b5)

The memory corruption error is a bit unexpected, when all I'm doing is just updating a quat. Are there limits in place on what `World.OriginOffset` can be that I'm violating? I guess multiplying the quats together gives some numerically unsound results and/or that can't be set for the reference space?

Is this even the best way to move a player around a scene?

This happens with both SK 0.3.9 and 0.4.0-preview.493, on Windows 10, with a Varjo XR-4. 

Edit: added remark on quat
Edit 2: origin mode is Floor, also happens with Local
Edit 3: enabled native debugging and added the call stack screenshot

**Comment by maluoi** [comment]
I would not recommend modifying OriginOffset in most cases! That one is exclusively for when you want to change the OpenXR reference space entirely, and I'd consider this to be a _very_ heavy handed action. Sending a bad quaternion to this could be problematic.

Renderer.CameraRoot is likely the thing you want to be working with here! It's designed for this type of manipulation, so you're likely to have a lot better success with it. I've got a small camera controller from a while back using this:
https://gist.github.com/maluoi/552f5b4a74deb9f4edd2007ff0ce0123

It's missing some important features, like it rotates around the center of the playspace instead of where the user is located, but it's a decent starting point!

**Comment by paulmelis** [comment]
Okay, that makes a lot more sense, when using `Renderer.CameraRoot` 😄
